gpu-compute,dev-hsa: ROCm 5.5+ support

abmerop opened the pull request:
ROCm 5.5 support including:
- Vendor packet completion signals
- Queue remapping race condition fix
- Backwards compatible GPR allocation
- Fix transient readBlob fatal reading kernel descriptor

Change-Id: Id0841407bec564c84f590c943f0609b17e01e14c

abmerop committed:
dev-amdgpu: Better handling for queue remapping

The amdgpu driver can, at *any* time, tell the device to unmap a queue
to force the queue descriptor to be written back to main memory in the
form of a memory queue descriptor (MQD). It will then immediately remap
the queue and continue writing the doorbell to the queue. It is possible
that the doorbell write occurs after the queue is unmapped but before it
is remapped. In this situation, we need to check the updated value of
the doorbell for the queue and write that to the queue after it is
mapped.

To handle this, a pending doorbell packet map is created to hold a
packet to replay when the queue is mapped. Because PCI in gem5
implements only the atomic protocol port, we cannot use the original
packet as it must respond in the same Tick. This patch fixes issues with
the doorbell maps not being cleared on unmapping to ensure the doorbell
is not found in writeDoorbell and places in the pending doorbell map.
This includes fixing the doorbell offset value in the doorbell to VMID
map which was is now multiplied by four as it is a dword address.

This was tested using tensorflow 2.0's MNIST example which was seeing
this issue consistently. With this patch it now makes progress and does
issue pending doorbell writes.

Change-Id: Ic6b401d3fe7fc46b7bcbf19a769cdea6814e7d1e

abmerop committed:
gpu-compute: Read dispatch packet with timing DMA

This fixes occasional readBlob fatals caused by the functional read of
system memory, seen often with the KVM CPU.

Change-Id: Ifccee666f62faa5b2fcf0a64a9d77c8cf95b3add

abmerop committed:
gpu-compute,dev-hsa: Send vendor packet completion signal

gem5 does not currently implement any vendor-specific HSA packets.
Starting in ROCm 5.5, vendor packets appear to end with a completion
signal. Not sending this completion causes gem5 to hang. Since these
packets are not documented anywhere and need to be reverse engineered we
send the completion signal, if non-zero, and finish the packet as is the
current behavior.

Testing: HIP examples working on most recent ROCm release (5.7.1).

Change-Id: Id0841407bec564c84f590c943f0609b17e01e14c

mattsinc commented on src/dev/hsa/hsa_packet.hh:
```diff
@@ -100,6 +100,13 @@ struct _hsa_barrier_or_packet_t
+    uint32_t padding[14];
+    uint64_t completion_signal;
+};
+static_assert(sizeof(_hsa_generic_vendor_pkt) == 64);
```
Do we need to document why 64?

abmerop replied:
I can change to Addr which should make it implicitly clear why.

mattsinc commented on src/gpu-compute/gpu_command_processor.cc:
```diff
@@ -473,18 +473,27 @@ GPUCommandProcessor::driver()
+ * completion signal, if any. However, in the future proper handing may be
+ * required for vendor specific packets.
+ *
+ * In the version of ROCm that is currently supported the runtime will send
```
Be explicit about version?

abmerop replied:
FS mode will work on anything 4.0-5.4 right now.  I kind of don't want to have to keep updating a comment. The one I replaced was still referencing ROCm 1.6.  IMO user can check documentation for what "supported" versions means

mattsinc replied:
Maybe we should just add a pointer to the documentation here then?

abmerop replied:
OK.  Doesn't seem to be any documentation though, so not sure what to point to right now

mattsinc replied:
I think the update is fine then, thanks.

abmerop committed:
gpu-compute: Update GPR allocation counts

GPR allocation is using fields in the AMD kernel code structure which
are not backwards compatible and are not populated in more recent
compiler versions. Use the granulated fields instead which is enfored to
be backwards compatible.

Change-Id: I718716226f5dbeb08369d5365d5e85b029027932

abmerop commented:
Adding a few more commits to this as they are somewhat related to support for latest ROCm.

mattsinc commented on src/dev/amdgpu/pm4_packet_processor.cc:
```diff
@@ -384,7 +384,7 @@ PM4PacketProcessor::mapQueues(PM4Queue *q, PM4MapQueues *pkt)
                 addr, pkt->mqdAddr, pkt->vmid, gpuDevice->lastVMID());
 
-        gpuDevice->mapDoorbellToVMID(pkt->doorbellOffset,
+        gpuDevice->mapDoorbellToVMID(pkt->doorbellOffset << 2,
```
If possible, it would be great to document why << 2

abmerop replied:
Added comment

mattsinc commented on src/dev/hsa/hsa_packet.hh:
```diff
@@ -100,6 +100,13 @@ struct _hsa_barrier_or_packet_t
+    uint32_t padding[14];
+    Addr completion_signal;
+};
+static_assert(sizeof(_hsa_generic_vendor_pkt) == 64);
```
Same thing here, it would be great if we could explain why 64.

abmerop replied:
Added comment

mattsinc reviewed (dismissed):
approving to at least run tests.  Some commenting/documentation questions though.

powerjg commented:
@mattsinc you don't have to approve to run the tests. They run as long as (1) it's not a draft and (2) the submitter has contributed to gem5 before.

By "approving" the PR you are allowing the PR to be merged *as is* without any changes.

mattsinc commented:
> @mattsinc you don't have to approve to run the tests. They run as long as (1) it's not a draft and (2) the submitter has contributed to gem5 before.
> 
> By "approving" the PR you are allowing the PR to be merged _as is_ without any changes.

I did not mean to press it, sorry.  I was trying to scroll and click on "Quote Reply", but my web browser was lagging and clicked this instead ... will try to avoid moving forward.  I did not see a way to undo once I'd clicked it.

powerjg commented:
np, that's why I asked :)

You can go back and select "request changes" to prevent merging, if you want.

![image](https://github.com/gem5/gem5/assets/109596/4623bab7-f708-47d5-8355-2852e4690994)
